src/target/riscv/riscv-013.c

```diff
@@ -1688,13 +1757,29 @@ static int examine(struct target *target)
 					info->index);
 			return ERROR_FAIL;
 		}
+		target->state = TARGET_HALTED;
+		target->debug_reason = DBG_REASON_DBGRQ;
+	}
+	/* FIXME: This is needed since register_read_direct relies on target->state
+	 * to work correctly, so, if target->state does not represent current state
+	 * of target, e.g. if a target is halted, but target->state is
+	 * TARGET_UNKNOWN, it can fail early, (e.g. accessing registers via program
+	 * buffer can not be done atomically on a running hart becuse mstatus can't
+	 * be prepared for a register access and then restored)
+	 * See https://github.com/riscv/riscv-openocd/pull/842#discussion_r1179414089
+	 */
+	const enum target_state saved_tgt_state = target->state;
+	const enum target_debug_reason saved_dbg_reason = target->debug_reason;
```

Comment:
In an earlier version of this merge request that I have reviewed, this hack to save & restore of the target state had been already removed and done differently (as [discussed here](https://github.com/riscv/riscv-openocd/pull/842/files#r1183104333)). I was happy with that version.

Now the save & restore is back. Can you please double-check if that is an intentional change. Or did the older code get back just by mistake?

Reply:
This is an intentional revert. Yesterday I've found a bug in prior implementation:
if `dcsr.cause == CSR_DCSR_CAUSE_TRIGGER`, calling `riscv013_halt_reason` will result in a call to `riscv_enumerate_triggers`, which is defined in `riscv.c` (universal for 0.11 and 0.13), so it uses `reg_cache` to determine size of `tselect`, and, since `reg_cache` is not initialized yet, gets `-1`:
```
Debug: 136 5 riscv.c:4494 riscv_get_register(): [riscv.cpu] No cache, reading tselect from target
Debug: 137 5 riscv-013.c:4178 riscv013_get_register(): [riscv.cpu] reading register tselect
Debug: 138 5 riscv-013.c:1529 register_read_direct(): [riscv.cpu] Reading tselect
Error: 139 5 riscv-013.c:868 access_register_command(): [riscv.cpu] -1-bit register tselect not supported.
```
I was going to explain this today, but didn't finish it in time for your review)

Reply:
Your [suggestion](https://github.com/riscv/riscv-openocd/pull/842#discussion_r1180436311) will probably work, but are no tests, either in `riscv-tests/debug` nor in our (at Syntacore) internal testing, which would explicitly target examine, so, I would like to stick with current solution, since it definitely won't change the behavior and introduce new bugs.

Reply:
Done. Link to the issue: #852

Reply:
Thank you.

> I would like for it to be assigned to me. Who should I ask to do so?

I was able to assign the issue to you.